experimental/SetPartitions/src/SetPartition.jl

```diff
@@ -0,0 +1,305 @@
+"""
+    SetPartition
+
+A set-partition is a partition of a set of upper and lower points into disjoint subsets. 
+Such partitions are often depicted as string diagrams where points in the same subset 
+are connected by lines. See also Section 4.1.1 in [Gro20](@cite).
+
+`SetPartition` stores two vectors `upper_points` and `lower_points`
+where same numbers correspond to points in the same subset. Both vectors are 
+automatically converted into normal form as returned by [`normal_form`](@ref).
+
+# Examples
+```jldoctest
+julia> SetPartition([2, 4], [4, 99])
+SetPartition([1, 2], [2, 3])
+```
+"""
+struct SetPartition <: AbstractPartition
+    upper_points::Vector{Int}
+    lower_points::Vector{Int}
+
+    function SetPartition(upper_points, lower_points)
+        (new_upper, new_lower) = _normal_form(convert(Vector{Int}, upper_points), 
+                                             convert(Vector{Int}, lower_points))
+        return new(new_upper, new_lower)
+    end
+end
+
+
+function hash(p::SetPartition, h::UInt)
+    return hash(p.upper_points, hash(p.lower_points, h))
+end
+
+function ==(p::SetPartition, q::SetPartition)
+    return p.lower_points == q.lower_points && p.upper_points == q.upper_points
+end
+
+function deepcopy(p::SetPartition)
+    return SetPartition(deepcopy(p.upper_points), deepcopy(p.lower_points))
+end
```

Comment:
Please not that `deepcopy` is indeed the method to call for a copy. When providing a method however, you need to define `deepcopy_internal` (see https://docs.julialang.org/en/v1/base/base/#Base.deepcopy).